### PR TITLE
Slice methods for byte array access

### PR DESCRIPTION
**Pull request by pettyjamesm**
Adds methods to Slice similar to `ByteBuffer::{hasArray, array, arrayOffset}` so that external code can more easily access the underlying byte array if one exists.

### CODE REVIEW

File: src/main/java/io/airlift/slice/Slice.java [hunk 1]
```diff
@@ -612,8 +636,8 @@ public void getBytes(int index, OutputStream out, int length)
     {
         checkIndexLength(index, length);
 
-        if (base instanceof byte[]) {
-            out.write((byte[]) base, (int) ((address - ARRAY_BYTE_BASE_OFFSET) + index), length);
+        if (hasBackingByteArray()) {
+            out.write(backingByteArray(), (backingByteArrayOffset() + index), length);
```
Comment (electrum) [inline src/main/java/io/airlift/slice/Slice.java]: Nit: no need for parenthesis

**Comment by pettyjamesm** [comment]
@electrum Removed the unnecessary parenthesis and renamed the methods. Added a javadoc comment mentioning the callers responsibility to check `hasByteArray()` and to use `byteArrayOffset()` when accessing the contents.

**Comment by pettyjamesm** [comment]
Not sure why the travis tests failed since the JVM crashed for the JDK 11 run. Previous test run completed successfully and the only change between the two was renaming methods and adding a javadoc comment.

**Comment by electrum** [comment]
Merged, thanks!

I re-ran the Travis build and it passed. Good investigation. It looks like the test is flakey given the limited memory of Travis. Maybe we can have it retry automatically: https://maven.apache.org/surefire/maven-surefire-plugin/examples/rerun-failing-tests.html

**Comment by pettyjamesm** [comment]
Honestly, I doubt that rerunning is an option here since the test container OOM killer is the culprit. If separating the next slice size from actually allocating that slice isn’t an option due to jvm max inlining depth defaults (which is a possibility) then maybe the test or test environment  needs to be reconsidered. I’d be worried about spurious failures plaguing the repo until the end of time otherwise

**Pull request merged**